lib/request-handler.js

```diff
@@ -4,469 +4,233 @@ var util = require('util');
- * @private
+ * Gets a connection from the next host according to the query plan or a NoHostAvailableError.
+ * @param {Function} callback
  */
-RequestHandler.prototype._getPooledConnection = function (host, callback) {
-  // Implementation detail: avoid the overhead of async flow control functions
-  var self = this;
-  host.borrowConnection(function (err, connection) {
-    if (err) {
-      return callback(err);
-    }
-    if (!self.client.keyspace || self.client.keyspace === connection.keyspace) {
-      // connection is ready
-      return callback(null, connection);
-    }
-    // switch keyspace
-    connection.changeKeyspace(self.client.keyspace, function (err) {
-      if (err) {
-        host.removeFromPool(connection);
-        return callback(err);
-      }
-      callback(null, connection);
-    });
-  });
+RequestHandler.prototype.getNextConnection = function (callback) {
+  RequestHandler.borrowNextConnection(
+    this._hostIterator, this.triedHosts, this.client.profileManager, this.client.keyspace, callback);
 };
 
 RequestHandler.prototype.log = utils.log;
 
 /**
  * Gets an available connection and sends the request
- * @param request
- * @param {QueryOptions|null} options
  * @param {Function} callback
  */
-RequestHandler.prototype.send = function (request, options, callback) {
-  if (this.request === null) {
-    options = options || utils.emptyObject;
-    if (options.captureStackTrace) {
-      Error.captureStackTrace(this.stackContainer = {});
-    }
+RequestHandler.prototype.send = function (callback) {
+  if (this.requestOptions.captureStackTrace) {
+    Error.captureStackTrace(this.stackContainer = {});
   }
-  this.request = request;
-  this.requestOptions = options;
   var self = this;
-  this._getNextConnection(options, function (err, c) {
+  this.loadBalancingPolicy.newQueryPlan(this.client.keyspace, this.requestOptions, function newPlanCb(err, iterator) {
     if (err) {
-      //No connection available
       return callback(err);
     }
-    self.connection = c;
-    self._sendOnConnection(request, options, callback);
+    self._hostIterator = iterator;
+    self._callback = callback;
+    self._startNewExecution();
   });
 };
 
-/**
- * Sends a request to the current connection, adapting the result and retrying, if necessary.
- * @param request
- * @param {QueryOptions} options
- * @param {Function} callback
- * @private
- */
-RequestHandler.prototype._sendOnConnection = function (request, options, callback) {
+RequestHandler.prototype._startNewExecution = function () {
+  var execution = new RequestExecution(this);
+  this._executions.push(execution);
   var self = this;
-  this.connection.sendStream(request, options, function readCallback(err, response) {
-    if (err) {
-      //Something bad happened, maybe retry or do something about it
-      return self._handleError(err, callback);
+  execution.start(function hostAcquired(host) {
+    if (!self.requestOptions.isIdempotent) {
```

Comment:
Shouldn't this only be checked after the 1st execution?

Reply:
The first execution has started, the check is made to schedule the next timeout.

Reply:
Ah I see,  I was following the code wrong.  Had interpreted it that this included the first execution, but if that is not the case then this makes sense.